proknow-sdk/Patient/Entities/StructureSet/StructureSetDraftLockRenewer.cs

```diff
@@ -0,0 +1,92 @@
+        private Timer _timer;
+        private bool _hasStarted;
+        private TimeSpan _lockRenewalBuffer;
+
+        /// <summary>
+        /// Creates a StructureSetDraftLockRenewer
+        /// </summary>
+        /// <param name="proKnow">Root object for interfacing with the ProKnow API</param>
+        /// <param name="structureSet">The structure set</param>
+        public StructureSetDraftLockRenewer(ProKnowApi proKnow, StructureSetItem structureSet)
+        {
+            _proKnow = proKnow;
+            _structureSet = structureSet;
+            _timer = null;
+            _hasStarted = false;
+            _lockRenewalBuffer = new TimeSpan(0, 0, proKnow.LockRenewalBuffer);
+        }
+
+        /// <summary>
+        /// Starts the timer
+        /// </summary>
+        public void Start()
+        {
+            if (!_hasStarted)
+            {
+                //DateTime expiresAt = DateTime.ParseExact(_structureSet.DraftLock.ExpiresAt, "yyyy-MM-dd HH:mm:ss,fff",
+                //    CultureInfo.InvariantCulture);
+                TimeSpan expiresIn = new TimeSpan(0, 0, 0, 0, _structureSet.DraftLock.ExpiresIn);
+                //TimeSpan period = (expiresAt - DateTime.UtcNow) - _lockRenewalBuffer;
+                TimeSpan period;
+                if (_lockRenewalBuffer < expiresIn)
+                {
+                    period = expiresIn - _lockRenewalBuffer;
+                }
+                else
+                {
+                    period = new TimeSpan(0);
+                }
+                _timer = new Timer(Run, null, new TimeSpan(0), period);
+                _hasStarted = true;
+            }
+        }
+
+        /// <summary>
+        /// Stops the timer
+        /// </summary>
+        public void Stop()
+        {
+            _timer.Dispose();
+            _hasStarted = false;
+        }
+
+        /// <summary>
+        /// The TimerCallback delegate that renews the draft lock
+        /// </summary>
+        private void Run(Object notUsed)
+        {
+            try
+            {
+                var json = _proKnow.Requestor.PutAsync($"/workspaces/{_structureSet.WorkspaceId}/structuresets/{_structureSet.Id}/draft/lock/{_structureSet.DraftLock.Id}").Result;
+                _structureSet.DraftLock = JsonSerializer.Deserialize<StructureSetDraftLock>(json);
+            }
+            catch (AggregateException ae)
+            {
+                foreach (var ex in ae.InnerExceptions)
+                {
+                    // Handle the situation where the lock was deleted while the timer was firing (e.g., during unit testing)
+                    if (!(ex is ProKnowHttpException) ||
+                        (ex.Message != "HttpError(Forbidden, Incorrect lock)" &&
+                         ex.Message != "HttpError(Forbidden, Structure set is not currently locked for editing)"))
+                    {
+                        throw ex;
+                    }

```

Comment:
Does this have the potential to suppress legitimate issues. For example, if a user makes their lock renewal buffer too short (e.g., 0), couldn't that result in the lock not being renewed properly?

Reply:
Instead of this approach, I moved the code in StructureSetItem to stop the draft lock renewer (dispose the timer) prior to asynchronous calls to the API to approve (commit) or discard the draft.  Additionally in StructureSetDraftLockRenewer, I added a check to make sure the timer still exists before attempting to renew it.  I believe that eliminates the legitimate scenarios where the timer might be executing the callback to renew when editing has already completed and the lock is no longer valid.